File: solutions/javascript/rail-fence-cipher/1/rail-fence-cipher.js
//
// This is only a SKELETON file for the 'Rail Fence Cipher' exercise. It's been provided as a
// convenience to get you started writing code faster.
//
export const encode = (input, rail) => {
  if (!input) return '';
  let spaceRemoved = input.replace(/\s+/g, '');
  let rows = Array(rail).fill('');
  let currentRow = 0;
  let goingDown = false;

  for (let char of spaceRemoved) {
    rows[currentRow] += char;

    if (currentRow === 0 || currentRow === rail - 1) {
      goingDown = !goingDown;
    }

    currentRow += goingDown ? 1 : -1;
  }

  return rows.join('');
};

export const decode = (cipher,rail) => {
   if (!cipher) return '';
  // Step 1: Build the zigzag pattern for each character position
  let pattern = [];
  let currentRow = 0;
  let goingDown = false;

  for (let i = 0; i < cipher.length; i++) {
    pattern.push(currentRow);
    if (currentRow === 0 || currentRow === rail - 1) {
      goingDown = !goingDown;
    }
    currentRow += goingDown ? 1 : -1;
  }

  
  let railLengths = Array(rail).fill(0);
  for (let i = 0; i < pattern.length; i++) {
    railLengths[pattern[i]]++;
  }


  let railsArray = [];
  let index = 0;
  for (let r = 0; r < rail; r++) {
    railsArray[r] = cipher.slice(index, index + railLengths[r]);
    index += railLengths[r];
  }


  let positions = Array(rail).fill(0);
  let decoded = '';

  for (let i = 0; i < pattern.length; i++) {
    let railIndex = pattern[i];
    decoded += railsArray[railIndex][positions[railIndex]];
    positions[railIndex]++;
  }

  return decoded;
};

File: solutions/python/rail-fence-cipher/1/rail_fence_cipher.py
def encode(message, rails):

    if rails == 1:
        return message

    rails_list = [""] * rails

    rail = 0
    direction = 1

    for char in message:
        rails_list[rail] += char

        rail += direction

        if rail == 0 or rail == rails - 1:
            direction *= -1

    return "".join(rails_list)

def decode(encoded_message, rails):

    if rails == 1:
        return encoded_message

    
    pattern = []
    rail = 0
    direction = 1

    for _ in encoded_message:
        pattern.append(rail)

        rail += direction
        if rail == 0 or rail == rails - 1:
            direction *= -1

    
    rail_counts = [pattern.count(r) for r in range(rails)]

  
    rails_list = []
    index = 0

    for count in rail_counts:
        rails_list.append(encoded_message[index:index + count])
        index += count

    rail_positions = [0] * rails
    result = ""

    for rail in pattern:
        result += rails_list[rail][rail_positions[rail]]
        rail_positions[rail] += 1

    return result

